### ISSUE TITLE
Tailscale kubernetes deployment is broken in prod with latest change (k8s)

### ISSUE THREAD

**Issue by opqpop** [tailscale/tailscale#6629]
### What is the issue?

Our Tailscale Kubernetes deployment has been working for the past few days, but has suddenly broken on its own w/o us having done anything. 

Error:

<img width="915" alt="image" src="https://user-images.githubusercontent.com/5566843/205791835-0394874d-b20f-4d66-a152-28e3b0ee8727.png">

"storing device ID in kube secret", "got non-200 status code 403"

There are a couple recent changes that happened ~28d ago from @danderson:
- https://github.com/tailscale/tailscale/blame/4b34c88426de324255f0906db8ae6ecb6da5d107/cmd/containerboot/kube.go#L72
- https://github.com/tailscale/tailscale/commit/bf2d3cd07471cb0501343f694594b3e0b0e3db18

Do you know if these might be related? 

Our k8 container (deployed on Google Kubernetes Engine):

```
        - name: ts-sidecar
          imagePullPolicy: Always
          image: "ghcr.io/tailscale/tailscale:latest"
          env:
            # Store the state in a k8s secret
            - name: TS_KUBE_SECRET
              value: "tailscale-auth-dracula"
            - name: TS_USERSPACE
              value: "false"
            - name: TS_AUTH_KEY
              valueFrom:
                secretKeyRef:
                  name: tailscale-auth-dracula
                  key: TS_AUTH_KEY
                  optional: true
          securityContext:  
            capabilities:
              add:
                - NET_ADMIN
```

What we've tried
- tried using `tailscale/tailscale:stable` docker image but still failed

Any thoughts on how we can further debug this? Ideally we can figure out which of the lines is raising this error, and even better, what `kubectl` commands we can run to repro the actual error to debug.

```

// storeDeviceID writes deviceID into the "device_id" data field of
// the kube secret secretName.
func storeDeviceID(ctx context.Context, secretName, deviceID string) error {
	// First check if the secret exists at all. Even if running on
	// kubernetes, we do not necessarily store state in a k8s secret.
	req, err := http.NewRequest("GET", fmt.Sprintf("/api/v1/namespaces/%s/secrets/%s", kubeNamespace, secretName), nil)
	if err != nil {
		return err
	}
	resp, err := doKubeRequest(ctx, req)
	if err != nil {
		if resp != nil && resp.StatusCode >= 400 && resp.StatusCode <= 499 {
			// Assume the secret doesn't exist, or we don't have
			// permission to access it.
			return nil
		}
		return err
	}

	m := map[string]map[string]string{
		"stringData": map[string]string{
			"device_id": deviceID,
		},
	}
	var b bytes.Buffer
	if err := json.NewEncoder(&b).Encode(m); err != nil {
		return err
	}
	req, err = http.NewRequest("PATCH", fmt.Sprintf("/api/v1/namespaces/%s/secrets/%s?fieldManager=tailscale-container", kubeNamespace, secretName), &b)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/strategic-merge-patch+json")
	if _, err := doKubeRequest(ctx, req); err != nil {
		return err
	}
	return nil
}
```

Thanks!

### OS

Linux

**Comment by DentonGentry** [comment]
To be clear, are you looking for:
- how to get the cluster back to functional ASAP? [v1.32.3](https://hub.docker.com/layers/tailscale/tailscale/v1.32.3/images/sha256-c5a3a05334beab8541b8582ac6b235334925bb73061881b095ba6dad4e97bf66?context=explore) was the prior stable release
- root cause analysis regarding why 1.34.0 is behaving differently and what we might do to address it (but not ASAP)

**Comment by danderson** [comment]
Quick fix: pin the container image to `:v1.32.3` instead of `:latest`, which just bumped to 1.34.0.

Figuring out a cause...

This does look like the changes we made to the container image PID1. According to the error, the tailscale pod doesn't have permissions to write to the TS_KUBE_SECRET, which... should be breaking a lot more than just writing the device ID, since that's also where tailscale's runtime state gets stored.

Can you check that the pod's service account has permissions to write to the `tailscale-auth-dracula `secret? Does `kubectl get -oyaml tailscale-auth-dracula` show that the secret contains several key/value pairs (e.g. keys named `_daemon` , `_machinekey`) ? Don't paste the values here since those are your node's cryptographic identity, I'm just curious to know if they exist at all.

What version of Kubernetes is your cluster running? The other possibility is that the patch method we're using to update the secret is maybe not supported in the version you're running, but afaik server side patching is several versions old by now.

**Comment by wolffiex** [comment]
Thanks so much for the quick reply! Here's our version info
```
ubectl version --output=yaml
clientVersion:
  buildDate: "2022-09-14T19:49:27Z"
  compiler: gc
  gitCommit: e4d4e1ab7cf1bf15273ef97303551b279f0920a9
  gitTreeState: clean
  gitVersion: v1.25.1
  goVersion: go1.19.1
  major: "1"
  minor: "25"
  platform: linux/amd64
kustomizeVersion: v4.5.7
serverVersion:
  buildDate: "2022-09-22T09:24:03Z"
  compiler: gc
  gitCommit: 4262ac74f84d8f8ec8f692ea2080483e932554f9
  gitTreeState: clean
  gitVersion: v1.22.15-gke.100
  goVersion: go1.16.15b7
  major: "1"
  minor: "22"
  platform: linux/amd64

WARNING: version difference between client (1.25) and server (1.22) exceeds the supported minor version skew of +/-1
```

**Comment by wolffiex** [comment]
We are having trouble verifying that the service account has permission to view the secret. 
```
kubectl get secret tailscale-auth-moonlit --as=dracula-account
Error from server (Forbidden): secrets "tailscale-auth-moonlit" is forbidden: User "dracula-account" cannot get resource "secrets" in API group "" in the namespace "default"
```

**Comment by wolffiex** [comment]
Ok, we can confirm that downgrading to the v1.32.3 fixes this issue.

**Comment by danderson** [comment]
I'll have to dig into the code for our kube storage backend, but I'm suspecting that we've just silently not been using that secret for state storage in your cluster this whole time due to permission errors, and the only change is that 1.34 made that a fatal error. I would expect to see a `Role` and `RoleBinding` in your cluster that grants "*" permissions to `dracula-account`, for this to work.

One possibility is that you might have permissions, but the Role might not include the `PATCH` verb, which is what the PID 1 uses. If that's the case, the fix would be to grant `*` verbs for the secret, since the pod is the owner of the secret and needs full read/write on it.

Let me know if you find anything regarding the permissions, if those are correct we may have some other issue. I'm looking at release notes now to confirm that server-side resource patching is expected to work,

**Comment by chantlong** [comment]
Downgrading to v.13.2.3 also fixed the issue for me.

**Comment by danderson** [comment]
What permissions are you granting to the pod for accessing the secret? Downgrading won't help in the long term unless we can work out what the issue is. So far it seems to be a permissions configuration problem, but without more details we can't root-cause.

**Comment by wolffiex** [comment]
Here's how we set this up
```
kubectl describe serviceaccount dracula-account
Name:                dracula-account
Namespace:           default
Labels:              <none>
Annotations:         <none>
Image pull secrets:  <none>
Mountable secrets:   dracula-account-token-ps4w7
Tokens:              dracula-account-token-ps4w7
Events:              <none>

kubectl describe rolebinding dracula-tailscale
Name:         dracula-tailscale
Labels:       <none>
Annotations:  <none>
Role:
  Kind:  Role
  Name:  tailscale
Subjects:
  Kind            Name             Namespace
  ----            ----             ---------
  ServiceAccount  dracula-account  
kubectl describe role tailscale
Name:         tailscale
Labels:       <none>
Annotations:  <none>
PolicyRule:
  Resources  Non-Resource URLs  Resource Names                   Verbs
  ---------  -----------------  --------------                   -----
  secrets    []                 []                               [create]
  secrets    []                 [tailscale-auth-codeslaw]        [get update]
  secrets    []                 [tailscale-auth-dracula]         [get update]
  secrets    []                 [tailscale-auth-moonlit-latest]  [get update]
  secrets    []                 [tailscale-auth-unisearch]       [get update]
```

**Comment by wolffiex** [comment]
> the Role might not include the PATCH verb
Looks like this may be the problem

**Comment by danderson** [comment]
Yup, that sounds right, thanks for investigating!

If you reapply the [role.yaml](https://github.com/tailscale/tailscale/blob/main/docs/k8s/role.yaml) or grant the patch permission in whatever way you deploy this, it should fix 1.34 for you.

I'll go and see if there's a clean way we can do concurrent updates without the patch primitive, to be more compatible with older configurations. Or at the very least, we can make the image fail with more obvious instructions on how to fix.

**Comment by wolffiex** [comment]
Well, thanks so much for the pointer! This appears to have been the problem. You guys rock!! Keep up the great work at tailscale, we love it.
